### PR TITLE
[Opik-263] update dataset to support flexible items format

### PR DESCRIPTION
**Pull request by alexkuzmik**
## Details
This PR removes dataset items format restriction making it possible to specify any item structure you need.
Since there is no longer pre-defined fields (input, expected_output, metadata), `DatasetItem` class has been removed from the `Dataset` public methods signatures in favor of regular dictionaries, it was also removed from `opik` namespace. DatasetItem is only being used under the hood as an implementation detail now.

The following code areas were impacted by the PR:
1. Dataset
2. Evaluation
3. Pytest integration
4. Fern generated code

An example in `examples` folder was updated, but **if there are cookbooks or documentation references to DatasetItem, they must be updated as well!**. FYI @jverre

## Testing
Some new unit and e2e tests were added or updated. But it requires manual testing as well.

### CODE REVIEW

File: sdks/python/src/opik/api_objects/dataset/dataset_item.py [hunk 1]
```diff
@@ -1,37 +1,46 @@
...
-            "metadata": self.metadata,
-        }
+    def get_content(self, include_id: bool = False) -> Dict[str, Any]:
+        content = {**self.model_extra}
+        if include_id:
+            content["id"] = self.id
+
+        return content
+
+    def content_hash(self) -> str:
```
Comment (japdubengsub) [inline sdks/python/src/opik/api_objects/dataset/dataset_item.py:37]: we could convert this method to `__hash__()` magick method for easier use, and (probably) cache this value with lru_cache/cached_property to speed up re-syncing process

**Review by japdubengsub — APPROVED** [review]

**Comment by japdubengsub** [comment]
Great!

File: sdks/python/src/opik/api_objects/dataset/dataset_item.py (hunk 1 above)
Comment (alexkuzmik) [reply sdks/python/src/opik/api_objects/dataset/dataset_item.py:37]: 1. Whenever we calculate the hash, it's always done on the fresh `DatasetItem` instance which hash function hasn't been called before, so at the moment we'll not benefit from caching (unless we make an external cash function for memorizing content dictionaries, which I think might be a bit of an overkill here at the moment). But it's a good idea in general, as soon as we have a case of a repeated hash calculation or we see that we have performance issues, we should probably get back to it.
2. I intentionally didn't name this method `__hash__()` to emphasize that it calculates hash only from the data that we consider a dataset item's content :) (id-like fields are not involved in the calculation, as you can see).

**Review by andrescrz — DISMISSED** [review]
I didn't go in depth into the tests nor the auto-generated Fern code. But the rest LGTM.

**Review by andrescrz — APPROVED** [review]

File: sdks/python/src/opik/evaluation/evaluator.py [hunk 2]
```diff
@@ -26,8 +26,8 @@ def evaluate(
     Args:
         dataset: An Opik dataset instance
 
-        task: A callable object that takes DatasetItem as input and returns
-            dictionary which will later be used for scoring
+        task: A callable object that takes dict with dataset item content
```
Comment (andrescrz) [inline sdks/python/src/opik/evaluation/evaluator.py:29]: Minor about wording, missing articles: 
- `..takes a dict with the dataset item content...` 
- `..returns a dic...`

**Review by andrescrz — APPROVED** [review]

**Review by ferc — CHANGES_REQUESTED** [review]
The `dataset.insert` should support any kind of values for the top-level keys, such us:
```python
dataset.insert(
    [
        {"input": "test", "some_number": 2, "some_array": ["foo", "bar"], "none_test": None}
    ]
)
```

That example is not working currently:
```pydantic.v1.error_wrappers.ValidationError: 4 validation errors for DatasetItem
data -> input
  value is not a valid dict (type=type_error.dict)
data -> some_number
  value is not a valid dict (type=type_error.dict)
data -> some_array
  value is not a valid dict (type=type_error.dict)
data -> none_test
  none is not an allowed value (type=type_error.none.not_allowed)```

**Review by ferc — APPROVED** [review]

**Pull request merged**